specs/experimental/gov-delegation.md

```diff
@@ -0,0 +1,440 @@
+
+## Backwards Compatibility
+
+The `GovernanceDelegation` contract ensures backwards compatibility by allowing the migration of delegation state from the
+`GovernanceToken`.
+
+## Migration
+
+All write functions in the `GovernanceDelegation` MUST check if the users interacting with it have been migrated by
+checking the `migrated` mapping from its [storage](#storage). If a user has not been migrated, the `GovernanceDelegation`
+MUST copy the delegation and checkpoint data from the token contract to its own state. After copying the data, the
+`GovernanceDelegation` MUST update the `migrated` mapping to reflect that the address has been migrated.
+
+## User Flow
+
+The following sections highlight the use cases that MUST be supported by the `GovernanceDelegation`, and the difference
+for basic delegations made from the `GovernanceToken` contract.
+
+### Partial Delegations
+
+Users MUST be able to perform partial delegations of their voting power to another address. This must be supported for
+both absolute and relative amounts.
+
+### Absolute & Relative Delegations
+
+Users MUST be able to delegate their voting power to another address with an absolute or relative amount.
+
+### Differences
+
+The main difference for delegations made from the `GovernanceToken` contract is that basic delegations are encapsulated in
+a [`Delegation`](#delegation) struct and forwarded to the `GovernanceDelegation` contract. Basic delegation can be achieved
+with a `Delegation` such as:
+
+```solidity
+Delegation({
+  delegatee: ...,
+  allowanceType: AllowanceType.Relative,
+  allowance: 10e4 // 100%
+})
+```
+
+The following diagram shows the sequence of a basic delegation performed from the `GovernanceToken` contract.
+
+```mermaid
+sequenceDiagram
+    User ->> GovernanceToken: call `delegate` or `delegateBySig`
+    GovernanceToken ->> GovernanceDelegation: call `delegateFromToken`
+    GovernanceDelegation -->> GovernanceToken: migrate if user hasn't been migrated
+    GovernanceDelegation ->> GovernanceDelegation: apply basic delegation
+```
+
+Once a user has been migrated to the `GovernanceDelegation`, the `GovernanceToken` MUST always use the `GovernanceDelegation`'s
+delegation state. The following diagram shows the control flow for this case.
+
+```mermaid
+flowchart TD
+    A1(User) -->|Calls delegation getter function | A2(GovernanceToken)
+    A2 --> A3{Is user migrated?}
+    A3 --> |Yes, fetch from GovernanceDelegation| A4(GovernanceDelegation)
+    A3 --> |No, use its state| A2
+    A2 --> |Return data| A1
+```
+
+## Delegation Validation
+
+When applying a new delegation set as part of the [`_delegate`](#_delegate) function, the `GovernanceDelegation` MUST check
+that:
+
+- The length of the `_newDelegations` array DOES NOT exceed `MAX_DELEGATIONS`.
+- The sum of `amount` in relative delegations DOES NOT exceed the `DENOMINATOR`, and each `amount` is greater than 0.
```

Comment:
Can you have both relative and absolute delegations in the same set? If so, what is the invariant that must hold globally?

Reply:
updated to clarify invariant:

> If a delegation set includes both absolute and relative delegations, the relative delegations MUST be applied first and the absolute delegations second with the remaining voting power. Therefore, the remaining voting power MUST be greater than or equal than the sum of the absolute delegations.